optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/util/PromiseUtils.java

```diff
@@ -16,11 +16,60 @@
 
 package org.optaplanner.openshift.employeerostering.gwtui.client.util;
 
+import javax.inject.Singleton;
+
+import com.google.gwt.core.client.GWT;
+import elemental2.core.Error;
+import elemental2.core.JsObject;
 import elemental2.promise.Promise;
+import elemental2.promise.Promise.CatchOnRejectedCallbackFn;
+import elemental2.promise.Promise.PromiseExecutorCallbackFn;
 
+@Singleton
 public class PromiseUtils {
 
-    public static Promise<Void> resolve() {
+    public Promise<Void> resolve() {
         return Promise.resolve((Promise<Void>) null);
     }
+
+    public <T> Promise<T> promise(PromiseExecutorCallbackFn<T> executor) {
+        return manage(new Promise<>(executor));
+    }
+
+    public <T> Promise<T> manage(Promise<T> promise) {
+        return promise.catch_(this.<T> getDefaultCatch());
+    }
+
+    public <V> CatchOnRejectedCallbackFn<? extends V> getDefaultCatch() {
+        return (e) -> {
+            if (e instanceof Throwable) {
+                GWT.getUncaughtExceptionHandler().onUncaughtException((Throwable) e);
+            } else if (e instanceof Error) {
+                Error error = (Error) e;
+                GWT.getUncaughtExceptionHandler().onUncaughtException(new Throwable(error.toString()));
+            } else {
+                if (null != e) {
+                    JsObject error = new JsObject(e);
+                    // Exceptions seems to have a reference to the actual exception stored in "__java$exception"
+                    // Hopefully this is also true in production mode, else most stack traces will be useless
+                    if (error.hasOwnProperty("__java$exception")) {
```

Comment:
Errai has a [utility class](https://github.com/errai/errai/blob/master/errai-common/src/main/java/org/jboss/errai/common/client/logging/util/StackTraceFormatter.java) that does this for the SLF4J logger in a way that gives translated stack traces in Google Chrome (provided source maps are available).

You might just want to use an SLF4J logger here and log the error so that Errai prints it to the console.

Reply:
https://github.com/kiegroup/optashift-employee-rostering/blob/2ebc8b5964a4eef7bb38598d917e4939af3b9d5a/optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/OptaShiftRosteringEntryPoint.java#L46-L66 does the stack trace translating. Unfortunately the error is usually not an exception object except if we explicitly throw the exception (which is usually never). Ideally, we would not need this method, but since Promises are native js, it seems GWT does not use `getUncaughtExceptionHandler` to report the exception. It does appear I might be able to simplify this logic: the stack that cause an exception is visible from the catch method, so I can just turn this to `GWT.getUncaughtExceptionHandler().onUncaughtException(new Throwable(e.toString()))`.